### PR TITLE
Fetches POA information from BGS and VACOLS and wires it through to the client

### PR DESCRIPTION
**Pull request by amprokop**
Fetch POA information from BGS and VACOLS.

We're using a new model, the PowerOfAttorney model, to manage all power of attorney actions. Hopefully this model will support future POA-related efforts!

Also, this introduces the PowerOfAttorneyMapper, a module that explicitly exists to convert data from other systems' format to ours. We're using it even in the fake PowerOfAttorneyRepository.

We're trying to move away from putting everything in the appeal repository, and we're also trying to make our classes as testable as possible. Hopefully these items help.

Testing:
- bgs and vacols rep information is successfully wired through to the client.
image

### CODE REVIEW

File: app/models/power_of_attorney.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  attr_accessor :bgs_representative_name,
+                :bgs_representative_type,
+                :vacols_id,
+                :file_number
+
+  def load_bgs_record!
+    result = bgs.fetch_poa_by_file_number(file_number)
+    self.bgs_representative_name = result[:representative_name]
+    self.bgs_representative_type = result[:representative_type]
```
Comment (joofsh) [inline app/models/power_of_attorney.rb]: minor nit but these `self.` are unnecessary

File: app/models/power_of_attorney.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    self
+  end
+
+  def overwrite_vacols_with_bgs_value
+    # case_record.bfso
+  end
+
+  def bgs
+    BGSService.new
```
Comment (joofsh) [inline app/models/power_of_attorney.rb]: since this is an instance method, it may be worth memoizing this `@bgs ||=`

File: app/services/external_api/bgs_service.rb [hunk 3]
```diff
@@ -33,6 +35,20 @@ def fetch_veteran_info(vbms_id)
       end
   end
 
+  def fetch_poa_by_file_number(file_number)
+    @poas[file_number] ||=
+      MetricsService.record("BGS: fetch veteran info for file number: #{file_number}",
+                            service: :bgs,
+                            name: "org.find_poas_by_file_number") do
+        bgs_poa = client.veteran.find_poas_by_file_number(file_number)
+        get_poa_from_bgs_poa(bgs_poa)
```
Comment (joofsh) [inline app/services/external_api/bgs_service.rb]: hmm, do we want to move this line outside the `MetricsService.record {}`? Based on the name/service you're passing, we're recording how long the request takes, but we're adding in additional ruby calculations which will throw off the data

File: app/services/power_of_attorney_repository.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
+class PowerOfAttorneyRepository
+  # :nocov:
+  include PowerOfAttorneyMapper
+  # returns either the data or false
+  def self.load_vacols_data(poa)
+    case_record = MetricsService.record("VACOLS: load_vacols_data #{poa.vacols_id}",
+                                        service: :vacols,
+                                        name: "load_vacols_data") do
```
Comment (joofsh) [inline app/services/power_of_attorney_repository.rb]: we should name this request something more specific, else it will get merged in with all other `load_vacols_data`. Maybe `PowerOfAttorneyRepository.load_vacols_data`

File: app/services/power_of_attorney_repository.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
+class PowerOfAttorneyRepository
```
Comment (joofsh) [inline app/services/power_of_attorney_repository.rb]: 👏 

Minor nit: Do we want this in `app/repositories`?

File: app/mappers/power_of_attorney_mapper.rb [hunk 6]
```diff
@@ -0,0 +1,68 @@
+module PowerOfAttorneyMapper
+  # This is here so when we include this module
+  # in classes (e.g. in PoaRepository),
+  # the class itself and not just its instance
+  # get the methods from this class.
+  def self.included(base)
+    base.extend(PowerOfAttorneyMapper)
+  end
```
Comment (amprokop) [inline app/mappers/power_of_attorney_mapper.rb]: Problem is we use this on both the class (the repository) and in class instances (bgs service). If we did `extend`, those methods wouldn't be available on class instances.

File: app/services/external_api/bgs_service.rb (hunk 3 above)
Comment (amprokop) [inline app/services/external_api/bgs_service.rb]: true enough, this should never take more than a few milliseconds but for the principle of the thing! i will move it out

File: app/models/power_of_attorney.rb (hunk 2 above)
Comment (amprokop) [inline app/models/power_of_attorney.rb]: i haven't much (or at all) looked into this, but do you think we generally over-memoize? this doesn't create a connection or make any external calls — what's the advantage of memoizing something that's not slow in the first place?

Comment (amprokop) [inline app/models/power_of_attorney.rb]: happy to memoize this for sure too

File: app/mappers/power_of_attorney_mapper.rb (hunk 6 above)
Comment (amprokop) [inline app/mappers/power_of_attorney_mapper.rb]: I'm not quite sure why we only use class methods on the Appeal Repository, but it seems out of scope to refactor that here. LMK if you disagree.

File: app/models/power_of_attorney.rb (hunk 1 above)
Comment (amprokop) [inline app/models/power_of_attorney.rb]: i think rubocop was complaining about this variable assignment cuz it was never used, will check

File: app/services/power_of_attorney_repository.rb (hunk 5 above)
Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: hmm i almost feel like that enshrines the difference between this and a "service". if you feel strongly about it i'm happy to go for it but i think it can live here too

File: app/services/power_of_attorney_repository.rb (hunk 4 above)
Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: cool done

File: app/models/power_of_attorney.rb (hunk 1 above)
Comment (joofsh) [inline app/models/power_of_attorney.rb]: ah okay, whatever makes rubocop happy then

File: app/services/power_of_attorney_repository.rb (hunk 5 above)
Comment (joofsh) [inline app/services/power_of_attorney_repository.rb]: I thought that was the pattern we agreed with when we were talking with Shane the other day. I added the new `HearingRepository` under `app/repositories`, but I can move it back

File: app/models/power_of_attorney.rb (hunk 2 above)
Comment (joofsh) [inline app/models/power_of_attorney.rb]: I don't think we over or under memoize. We sometimes _incorrectly_ memoize on the class rather than an instance. Because `BGSService` makes API calls I felt it make sense to memoize. We had bugs in dispatch specifically because we didn't memoize `BGSService.new`. But I don't know the context here as well, and don't feel too strongly about it

File: lib/fakes/bgs_service.rb [hunk 7]
```diff
@@ -1,7 +1,10 @@
 class Fakes::BGSService
+  include PowerOfAttorneyMapper
```
Comment (joofsh) [inline lib/fakes/bgs_service.rb]: Cool 👍

**Comment by kierachell** [comment]
@amprokop 

### From UAT right now:
```
[2017-05-23T18:46:20.458650 #2769] DEBUG -- : <?xml version='1.0' encoding='UTF-8'?><S:Envelope xmlns:S="http://schemas.xmlsoap.org/soap/envelope/"><S:Header><work:WorkContext xmlns:work="http://oracle.com/weblogic/soap/workarea/">rO0ABXdUAB13ZWJsb2dpYy5hcHAuQ29ycG9yYXRlRGF0YUVBUgAAANYAAAAjd2VibG9naWMud29ya2FyZWEuU3RyaW5nV29ya0NvbnRleHQABjEwLjIuMAAA</work:WorkContext></S:Header><S:Body><ns2:findPOAsByFileNumbersResponse xmlns:ns2="http://org.services.vetsnet.vba.va.gov/"><return><fileNumber>516517690S</fileNumber><message>No PERSON found for 516517690S</message></return></ns2:findPOAsByFileNumbersResponse></S:Body></S:Envelope>
```
Seems like you're passing `BFCORLID` without stripping the `S` or `C`

### With proper file number:
```
[6] pry(main)> bgs.org.find_poas_by_file_number("516517690S")
=> {:file_number=>"516517690S", :message=>"No PERSON found for 516517690S"}
[7] pry(main)> bgs.org.find_poas_by_file_number("516517690")
=> {:file_number=>"516517690",
 :power_of_attorney=>
  {:legacy_poa_cd=>"095",
   :nm=>"ITALIAN AMERICAN WAR VETERANS OF THE US, INC.",
   :org_type_nm=>"POA National Organization",
   :ptcpnt_id=>"45994"},
 :ptcpnt_id=>"600161064"}
```

**Comment by amprokop** [comment]
Awesome! Actually, @sharonwarner already fixed this in https://github.com/department-of-veterans-affairs/caseflow/pull/2125/files. Let's test again when merged!

File: app/services/power_of_attorney_repository.rb (hunk 5 above)
Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: ah sorry i missed this comment — either way is fine with me
